internal/core/textchange.go

```diff
@@ -0,0 +1,10 @@
+package core
+
+type TextChange struct {
+	TextRange
+	NewText string
+}
+
+func (t TextChange) ApplyTo(text string) string {
+	return text[:t.Pos()] + t.NewText + text[t.End():]
```

Comment:
Yeah, in JS this is fast because ropes, but in Go this is super super slow. I think we're going to have to batch these. I haven't read onward from here but hopefully the formatter doesn't heavily depend on being able to repeatedly apply edits...